### PR TITLE
Add relabelling to target management.

### PR DESCRIPTION
**Pull request by fabxc**
This commit adds a relabelling stage on the set of base
labels from which a target is created. It allows to drop
targets and rewrite any regular or internal label.

### CODE REVIEW

File: config/config.proto [hunk 1]
```diff
@@ -58,9 +58,32 @@ message DNSConfig {
 	optional string refresh_interval = 2 [default = "30s"];
 }
 
+// The configuration for relabelling of target label sets.
+message RelabelConfig {
+	// A list of labels from which values are taken and concatenated
+	// with separator in order.
```
Comment (juliusv) [inline config/config.proto]: s/separator/the configured separator/

File: retrieval/discovery/dns.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ const (
 	resolvConf = "/etc/resolv.conf"
 
 	dnsSourcePrefix = "dns"
+	DNSNameLabel    = clientmodel.MetaLabelPrefix + "dns_name"
```
Comment (juliusv) [inline retrieval/discovery/dns.go]: `dns_srv_name` to avoid ambiguity with the instance CNAMEs that the SRV record resolves to.

**Comment by brian-brazil** [comment]
:+1:

File: retrieval/target.go [hunk 3]
```diff
@@ -258,7 +258,9 @@ func (t *target) RunScraper(sampleAppender storage.SampleAppender) {
 	t.Lock() // Writing t.lastScrape requires the lock.
 	t.lastScrape = time.Now()
 	t.Unlock()
-	t.scrape(sampleAppender)
+	if err := t.scrape(sampleAppender); err != nil {
```
Comment (juliusv) [inline retrieval/target.go]: Remove this as discussed.

File: retrieval/target.go [hunk 4]
```diff
@@ -293,7 +295,9 @@ func (t *target) RunScraper(sampleAppender storage.SampleAppender) {
 				targetIntervalLength.WithLabelValues(intervalStr).Observe(
 					float64(took) / float64(time.Second), // Sub-second precision.
 				)
-				t.scrape(sampleAppender)
+				if err := t.scrape(sampleAppender); err != nil {
```
Comment (juliusv) [inline retrieval/target.go]: dito

File: retrieval/targetmanager.go [hunk 5]
```diff
@@ -326,11 +326,19 @@ func (tm *TargetManager) targetsFromGroup(tg *config.TargetGroup, cfg *config.Sc
...
 
-		address, ok := labels[clientmodel.AddressLabel]
-		if !ok {
+		if _, ok := labels[clientmodel.AddressLabel]; !ok {
 			return nil, fmt.Errorf("Instance %d in target group %s has no address", i, tg)
 		}
 
+		labels, err := Relabel(labels, cfg.RelabelConfigs()...)
+		if err != nil {
+			return nil, fmt.Errorf("Error on relabelling instance %d in target group %s: %s", i, tg, err)
```
Comment (juliusv) [inline retrieval/targetmanager.go]: s/Error/error/

s/on/while/

File: retrieval/relabel.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	return out, nil
+}
+
+func relabel(labels clientmodel.LabelSet, cfg *config.RelabelConfig) (clientmodel.LabelSet, error) {
+	pat, err := regexp.Compile(cfg.GetRegex())
+	if err != nil {
+		return nil, err
+	}
+
+	var values []string
```
Comment (juliusv) [inline retrieval/relabel.go]: maybe `make` this with a capacity of `len(cfg.GetSourceLabel())` upfront?

**Comment by juliusv** [comment]
:+1: besides comments. Nice!

**Comment by fabxc** [comment]
All done.

**Pull request merged**